docs/proposals/parameterize-strategies.md

```diff
@@ -0,0 +1,267 @@
+
+### Goals
+
+- Introduce a `spec.params` field across Build and BuildRun API´s, as a way for users to provide N parameters according to their preferred strategy of choice. Defining `spec.params` in a BuildRun overrides any definition of the same parameter in Build´s.
+
+- Introduce a `spec.params`on Strategies, both cluster or namespaced scope. This allow strategy administrators to layout the definition of N parameters, as required for their strategies.
+
+- Revisit what we consider today as default Parameters that are defined on runtime when creating `TaskRuns`. These are `DOCKERFILE`, `CONTEXT_DIR` and `BUILDER_IMAGE`. Decide if we need to reduce or extend the default definitions.
+
+- Expose the usage of Tekton Params directly on the strategies, in favor of simplicity and readability. This boils down to `inputs.params.DOCKERFILE`, instead of `build.dockerfile`.
+
+### Non-Goals
+
+- Build from scratch a paremeterize mechanism, by replacing the one available in Task/TaskRuns.
+
+- Provide support for Tekton Params of the type `array`.
+
+## Proposal
+
+### Part 1: Introduce spec.params
+
+Introduce the `spec.params` field across Build, BuildRun and BuildStrategies resources. This will define a list  of parameters definition, of the type `string`. This new `spec.params` does not provide support for Tekton params of the type `array`. This field can only be use, if the parameter is supported in the strategy of choice.
+
+For example:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: ClusterBuildStrategy
+metadata:
+  name: a-cluster-strategy
+spec:
+  buildSteps: #Content omitted for this example
+  params:
+  - name: a-param
+    description: A description of this parameter definition.
+    default: The default parameter value.
+```
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+metadata:
+  name: a-build
+spec:
+  source: #Content omitted for this example
+  strategy:
+    name: a-strategy-with-params
+  output: #Content omitted for this example
+  params:
+  - name: a-param
+    value: A parameter value.
+```
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata:
+  name: a-buildrun
+spec:
+  buildRef:
+    name: a-build
+  params:
+  - name: a-param
+    value: Another parameter value because my build is not up-to-date.
+```
+
+_Note_: If a **Buildrun** specifies `a-param` via its `spec.params` ,this will override the value defined in the `a-build`. In other words, BuildRuns have a higher priority when defining parameters.
+
+### Part 2: Runtime Parameters
+
+As mentioned in the Goals section, we currently define three parameters that can be constructed on runtime, also know as _runtime parameters_. Runtime parameters are define during runtime, on the creation of a Taskrun, where the paremeter definition and the parameter value mapping takes place. These runtime parameters are:
+
+- DOCKERFILE
```

Comment:
Added a very generic comment because I forgot on this one (_deprecation_):

> This EP also propose to stop having an `spec.dockerfile` API field, but rather to move its definition under the new `spec.params`.

Reply:
I'm actually more inclined to keep `spec.Dockerfile` since it is a popular container image build concept. I would make that conversation a non-goal for this EP.

Reply:
Because I see a lack of consensus here and because our main goal should be to provide the flexibility of parameterize strategies. I will move this discussion into https://github.com/shipwright-io/build/issues/411 or a new issue. For the time being and opting keeping this on the same way it is, to avoid blocking this PR.